api/API_VERSIONING.md

```diff
@@ -0,0 +1,203 @@
+# API versioning guidelines
+
+The Envoy project (and in the future [UDPA](https://github.com/cncf/udpa)) takes API stability and
+versioning seriously. Providing stable APIs is a necessary step in ensuring API adoption and success
+of the ecosystem. Below we articulate the API versioning guidelines that aim to deliver this
+stability.
+
+# API semantic versioning
+
+The Envoy APIs consist of a family of packages, e.g. `envoy.admin.v2alpha`,
+`envoy.service.trace.v2`. Each package is independently versioned with a protobuf semantic
+versioning scheme based on https://cloud.google.com/apis/design/versioning.
+
+The major version for a package is captured in its name (and directory structure). E.g. version 2
+of the tracing API package is named `envoy.service.trace.v2` and its constituent protos are located
+in `api/envoy/service/trace/v2`. Every protobuf must live directly in a versioned package namespace,
+we do not allow subpackages such as `envoy.service.trace.v2.somethingelse`.
+
+Minor and patch versions will be implemented in the future as described in
```

Comment:
It's a little hard to find this specific information in that bug and the attached docs.  It might be a good idea to spell out the details here.

Reply:
+1

Reply:
I really don't want to get too prescriptive on how this works yet, as it's a detail that doesn't affect the v3 major version cut in Envoy 1.12.0. I have opened https://github.com/envoyproxy/envoy/issues/8416 and will link.